python/tank_vendor/shotgun_deploy/import_handler.py

```diff
@@ -0,0 +1,323 @@
+                # couldn't be found. not valid python
+                continue
+
+            # add the importable name to the list of top-level namespaces
+            self._namespaces.append(name)
+
+        if not self._namespaces:
+            # oops, no top-level namespaces found. this could indicate a bad
+            # core path.
+            raise RuntimeError(
+                "Could not find any top-level core namespaces. This may could "
+                "mean that the supplied core path (%s) is not valid" % (path,)
+            )
+
+        # acquire a lock to prevent issues with other threads importing at the
+        # same time.
+        imp.acquire_lock()
+
+        # sort by package depth, deeper modules first
+        module_names = sorted(
+            sys.modules.keys(),
+            key=lambda module_name: module_name.count("."),
+            reverse=True
+        )
+
+        for module_name in module_names:
+
+            # just to be safe, don't re-import this module.
+            # we always use the first one added to `sys.meta_path` anyway.
+            if module_name == __name__:
+                continue
+
+            # extract just the package name
+            pkg_name = module_name.split(".")[0]
+
+            if pkg_name in self._namespaces:
+                # the package name is in one of the new core namespaces. we
+                # delete it from sys.modules so that the custom import can run.
+                del sys.modules[module_name]
+
+        # clear out the previously found module info
+        self._module_info = {}
+
+        # release the lock so that other threads can continue importing from
+        # the new core location.
+        imp.release_lock()
+
+    @property
+    def core_path(self):
+        """The core_path for this importer.
+
+        :returns: str path to the core being used for imports
+        """
+        return self._core_path
+
+    @property
+    def namespaces(self):
+        """The namespaces this importer operates on.
+
+        :returns: a list where each item is a namespace str
+        """
+        return self._namespaces
+
+def swap_core(core_path):
+    """Swap the current core with the core located at the supplied path
+
+    Actually just unloads the existing core and ensures an import handler
+    exists that points to the supplied core path. When this method completes,
+    all core namespaces will be removed from `sys.modules`. The required
+    modules will need to be reimported after this method is called.
+
+    :param core_path: The path to the new core to use upon import.
+
+    """
+
+    # see if there's already a core import handler in use
+    import_handler = None
+    for handler in sys.meta_path:
+        if isinstance(handler, CoreImportHandler):
+            import_handler = handler
+            break
+
+    if not import_handler:
+        # no import handler yet, create a new one
+        import_handler = CoreImportHandler(core_path)
```

Comment:
does it make sense to move this into the `__init__` of the `tank` module itself rather than having it start up automatically the first time?